tests/test_cli.py

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+
+from argparse import ArgumentParser
+
+from .common import test_result_collector as trc
+
+from model_analyzer.config.input.config_command_profile import ConfigCommandProfile
+from model_analyzer.config.input.config_command_analyze import ConfigCommandAnalyze
+from model_analyzer.config.input.config_command_report import ConfigCommandReport
+from model_analyzer.cli.cli import CLI
+
+
+class ArgumentParserSubclass(ArgumentParser):
+
+    def print_help(self, file=None):
+        super().print_help(file)
+        self._did_print_help = True
+
+    _did_print_help = False
+
+
+class CLISubclass(CLI):
+
+    def __init__(self):
```

Comment:
Instead of re-implementing the function, I think you should use super().__init__():

```
def __init__(self):
    super().__init__()
    self._parser = ArgumentParserSubclass()
```

Reply:
this doesn't work

```
Traceback (most recent call last):
  File "/opt/triton-model-analyzer/tests/test_cli.py", line 88, in test_help_message_no_args
    self.assertRaises(SystemExit, cli.parse)
  File "/usr/lib/python3.8/unittest/case.py", line 816, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/usr/lib/python3.8/unittest/case.py", line 202, in handle
    callable_obj(*args, **kwargs)
  File "/opt/triton-model-analyzer/model_analyzer/cli/cli.py", line 137, in parse
    if args.subcommand is None:
AttributeError: 'Namespace' object has no attribute 'subcommand'
```

if i call `super().__init__()`, it will already create a regular ArgumentParser object and do things with it. by the time i can reassign the _parser data member to my own custom subclass, it's too late, and a future call to `cli.parse()` have its state messed up

Reply:
You're right :(.  Hardcoded class inside of constructor is tough to work around.

Reply:
As mentioned on the phone, lets try to use mock/patch for this. I think it fits in seamlessly and results in both of your subclasses going away

Reply:
switched to using [unittest.mock](https://docs.python.org/3/library/unittest.mock.html) 👍